st3/sublime_lib/resource_path.py

```diff
@@ -11,18 +11,102 @@
 __all__ = ['ResourcePath']
 
 
-def get_resource_roots():
-    return {
-        'Packages': sublime.packages_path(),
-        'Cache': sublime.cache_path(),
-    }
+def _abs_parts(path):
+    if path.root:
+        return (path.drive, path.root) + path.parts[1:]
+    else:
+        return path.parts
+
+def _file_relative_to(path, base):
+    """
+    Like Path.relative_to, except:
+
+    - `base` must be a single Path object.
+    - The error message is blank.
+    - Only a tuple of parts is returned.
+
+    Surprisingly, this is much, much faster.
+    """
+    child_parts = _abs_parts(path)
+    base_parts = _abs_parts(base)
+
+    n = len(base_parts)
+    cf = path._flavour.casefold_parts
+
+    if n == 0:
+        compare = (path.root or path.drive)
+    else:
+        compare = cf(child_parts[:n])
+
+    if compare != cf(base_parts):
+        raise ValueError()
+
+    return child_parts[n:]
+
+
+class ResourceRoot():
+    def __init__(self, root, path):
+        self.resource_root = ResourcePath(root)
+        self.file_root = Path(path)
+
+    def resource_to_file_path(self, resource_path):
+        resource_path = ResourcePath(resource_path)
+
+        parts = resource_path.relative_to(self.resource_root)
+        if parts == ():
+            return self.file_root
+        else:
+            return self._package_file_path(*parts)
+
+    def file_to_resource_path(self, file_path):
+        file_path = wrap_path(file_path)
+
+        if not file_path.is_absolute():
+            raise ValueError("Cannot convert a relative file path to a resource path.")
+
+        parts = _file_relative_to(file_path, self.file_root)
+        # parts = file_path.relative_to(self.file_root).parts
+        if parts == ():
```

Comment:
And this check for `None` and return that.

Reply:
Done.